Helm logging support (#1)

rockb1017 opened the pull request:
Improving the helm chart:
- k8s tagger to enrich with k8s metadata from container logs
- remove hostport from default values.yaml
- Optimize container log parsing by specifying the container runtime
- make it configurable to collect otel collector log
- create cluster role and role binding required for k8s metadata enrichment

sumo-drosiek commented on charts/opentelemetry-collector/templates/_config.tpl:
```diff
@@ -121,3 +122,116 @@ service:
+          - remove: uid
+{{- if .Values.agentCollector.containerLogs.enrichK8sMetadata }}
+processors:
+  k8s_tagger:
```
`groupbyattrs` processot is required to use before `k8s_tagger` as `k8s_tagger` uses resource labels

```
  groupbyattrs:
    keys:
      - k8s.container.name
      - k8s.namespace.name
      - k8s.pod.name
      - run_id
      - k8s.pod.uid
```

tigrannajaryan replied:
`groupbyattrs` will have a performance impact. Probably we have to do it now, but for the future I think it would be better if `filelog` allowed placing some attributes directly into the resource and we use that capability in k8s collection. I opened an issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2725

rockb1017 replied:
I am putting it in resource label, no? 
```
      - type: metadata
        attributes:
          stream: 'EXPR($.stream)'
          k8s.container.name: 'EXPR($.container_name)'
          k8s.namespace.name: 'EXPR($.namespace)'
          k8s.pod.name: 'EXPR($.pod_name)'
          run_id: 'EXPR($.run_id)'
          k8s.pod.uid: 'EXPR($.uid)'
        resource:
          k8s.pod.uid: 'EXPR($.uid)'
```

rockb1017 replied:
I confirm enrichment works without using groupbyattrs. 
```
2021-03-20T00:05:40.172Z        INFO    loggingexporter/logging_exporter.go:471 LogsExporter    {"#logs": 1}
2021-03-20T00:05:40.173Z        DEBUG   loggingexporter/logging_exporter.go:498 ResourceLog #0
Resource labels:
     -> k8spoduid: STRING(eda0a956-99ec-49cb-a954-d733e2f881dd)
     -> k8s.pod.annotations.splunk.com/index: STRING(test)
     -> k8s.pod.name: STRING(loggen-zjklf)
     -> k8s.namespace.name: STRING(default)
     -> k8s.pod.startTime: STRING(2021-03-20 00:03:32 +0000 UTC)
     -> k8s.pod.uid: STRING(eda0a956-99ec-49cb-a954-d733e2f881dd)
     -> k8s.node.name: STRING(ip-192-168-82-91.us-west-1.compute.internal)
     -> k8s.pod.labels.hello: STRING(loggen)
     -> k8s.pod.labels.app: STRING(loggen)
InstrumentationLibraryLogs #0
InstrumentationLibrary  
LogRecord #0
Timestamp: 1616198740006040006
Severity: Undefined
ShortName: 
Body: {
     -> log: STRING(num: 127 | loggen-zjklf | 2021-03-20T00:05:40.005870 | rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr
)
}
Attributes:
     -> k8s.pod.uid: STRING(eda0a956-99ec-49cb-a954-d733e2f881dd)
     -> run_id: STRING(0)
     -> stream: STRING(stdout)
     -> k8s.container.name: STRING(loggen)
     -> k8s.namespace.name: STRING(default)
     -> k8s.pod.name: STRING(loggen-zjklf)
```

@tigrannajaryan 
However, when i ingest this log to splunk, I only see metadata in `Attributes`, not in `Resources`
![image](https://user-images.githubusercontent.com/12387289/111853046-8de66e80-88d6-11eb-9971-3385c49ee1fb.png)
What should I do to ingest data in `Attributes` ?

tigrannajaryan replied:
@rockb1017 see if @atoulme can help, I have never used the HEC exporter. My guess is that HEC exporter does not handle Resource attributes, but I may be wrong. If that's the case then it is a bug in HEC exporter.

tigrannajaryan commented:
@rockb1017 please sign the CLA.

rockb1017 commented:
> @rockb1017 please sign the CLA.

Do I do a individual CLA?

tigrannajaryan commented:
@rockb1017 is this PR still needed?

rockb1017 commented:
@tigrannajaryan 
I rebased and updated my commit. 
It includes 
* k8s tagger to enrich with k8s metadata from container logs
* remove hostport from default values.yaml
* Optimize container log parsing by specifying the container runtime
* make it configurable to collect otel collector log
* create cluster role and role binding required for k8s metadata enrichment
* disable liveness and readiness because it is not working atm

sumo-drosiek commented on charts/opentelemetry-collector/templates/clusterRole.yaml:
```diff
@@ -0,0 +1,24 @@
+    - "get"
+    - "watch"
+{{- end -}}
+{{- end -}}
```
please add newline at the EOF

rockb1017 replied:
thanks

sumo-drosiek commented on charts/opentelemetry-collector/templates/clusterRoleBinding.yaml:
```diff
@@ -0,0 +1,18 @@
+  name: {{ include "opentelemetry-collector.serviceAccountName" . }}
+  namespace: {{ .Release.Namespace }}
+{{- end -}}
+{{- end -}}
```
please add newline at the EOF

tigrannajaryan commented:
EasyCLA is stuck. Let me try to close and reopen.

tigrannajaryan commented:
@rockb1017 please update PR description / commit message to reflect what you listed here https://github.com/open-telemetry/opentelemetry-helm-charts/pull/38#issuecomment-816338812

naseemkullah commented on charts/opentelemetry-collector/values.yaml:
```diff
@@ -110,25 +110,21 @@ ports:
     enabled: true
     containerPort: 4317
     servicePort: 4317
-    hostPort: 4317
```
Can you please keep the hostPorts enabled conditionally if being run as a daemonset (agent) ?

rockb1017 replied:
@naseemkullah Could you explain why you would want it by default? 
without using hostPort, it can be accessed through using k8s service. 
in my AWS EKS cluster, it fails to be deployed when trying to use hostPort. it seems that it requires additional cluster configuration to map it to host port. It would make it a bit more complicated to deploy it.

naseemkullah replied:
This how the instrumented workloads communicate with the collector when it runs as a daemon set in agent mode, I agree that in standalone collector mode they should be disabled and the k8s service used, hence enabling them under the condition that agent mode is enabled.

rockb1017 replied:
so pods that are running apps that has been instrumented to send trace information are sending it to its node ip it is running on?

naseemkullah replied:
thats right! with something like this in their env:

```
      - name: OTEL_EXPORTER_JAEGER_AGENT_HOST
        valueFrom:
          fieldRef:
            fieldPath: status.hostIP
 ```

naseemkullah replied:
the alternative is running the otel-collector agent as a sidecar container, which is outside the scope of this helm chart

dmitryax replied:
I agree, let's keep host ports. The only stable pipeline of the otel collector currently is tracing. And running agent with no ports exposed doesn't provide any value for traces collection.

dmitryax replied:
> in my AWS EKS cluster, it fails to be deployed when trying to use hostPort. it seems that it requires additional cluster configuration to map it to host port. It would make it a bit more complicated to deploy it.

@rockb1017 Could you please share the errors you see? I never saw any issues with exposed ports on this deamonset.

naseemkullah commented on charts/opentelemetry-collector/templates/_pod.tpl:
```diff
@@ -37,17 +37,22 @@ containers:
-      httpGet:
-        path: /
-        port: 13133
+    # livenessProbe: # TODO: does not work. it keeps killing pod. 
+    #   httpGet:
+    #     path: /
+    #     port: 13133
+    # readinessProbe:
+    #   httpGet:
+    #     path: /
+    #     port: 13133
```
This works prior to this commit, do you know what might have affected it? Would be nice to sort out before merging.

rockb1017 replied:
Hmm I just tested it and it is working again. I will uncomment this part. Thanks!

dmitryax commented on charts/opentelemetry-collector/values.yaml:
```diff
@@ -137,6 +133,17 @@ agentCollector:
+    includeAgentLogs: false
+    enrichK8sMetadata: true
+    listOfAnnotations: []
+    listOfLabels: []
```
K8s metadata applicable not only to logs, but widely used for traces and metrics as well.

All the default configuration can be easily overridden (by merging with defaults) in user's values.yaml. So I don't see a lot of benefit in providing extra configuration ways. Especially `listOfAnnotations` and `listOfLabels` can be easily replaced by the existing configuration method.
```
config:
  processors:
    k8s_tagger:
      annotations: ...
      labels: ...
```
it's just one line more to print than
```
agentCollector:
  containerLogs:
    listOfAnnotations:
    listOfLabels:
```
but less configuration options -> less confusion.

Speaking of `enrichK8sMetadata`, I would recommend to move the option one level up to be able to apply it to any of the pipelines. Like:
```
agentCollector:
  pipelinesWithK8sMetadata: [metrics, traces, logs]
```

Let's keep `agentCollector.k8sMetadataPipelines: []` by default.

@tigrannajaryan @naseemkullah thoughts?

naseemkullah replied:
> K8s metadata applicable not only to logs, but widely used for traces and metrics as well.
> 
> All the default configuration can be easily overridden (by merging with defaults) in user's values.yaml. So I don't see a lot of benefit in providing extra configuration ways. Especially `listOfAnnotations` and `listOfLabels` can be easily replaced by the existing configuration method.
> 
> ```
> config:
>   processors:
>     k8s_tagger:
>       annotations: ...
>       labels: ...
> ```
> 
> it's just one line more to print than
> 
> ```
> agentCollector:
>   containerLogs:
>     listOfAnnotations:
>     listOfLabels:
> ```
> 
> but less configuration options -> less confusion.

Good catch 💯 

> Speaking of `enrichK8sMetadata`, I would recommend to move the option one level up to be able to apply it to any of the pipelines. Like:
> 
> ```
> agentCollector:
>   pipelinesWithK8sMetadata: [metrics, traces, logs]
> ```
> 
> Let's keep `agentCollector.k8sMetadataPipelines: []` by default.
> 

I just learned about this. so we would have both `agentCollector.pipelines` and `agentCollector.pipelinesWithK8sMetadata`?

Seeing that this is a helm chart maybe we just add the `k8s_tagger` and so on directly in the pipelines out of the box if that makes sense?

dmitryax replied:
> I just learned about this. so we would have both agentCollector.pipelines and agentCollector.pipelinesWithK8sMetadata?

right

> Seeing that this is a helm chart maybe we just add the k8s_tagger and so on directly in the pipelines out of the box if that makes sense?

Sounds good, but having `k8s_tagger` enabled by default requires installing the cluster role. At least we need a way to opt out from that.

tigrannajaryan replied:
> but less configuration options -> less confusion.

I agree.

rockb1017 replied:
I think by having less config, you are making the users to study more on documentation to properly set up configOverride. 
For example, 
```
      k8s_tagger:
        passthrough: false
        auth_type: "kubeConfig"
        pod_association:
          - from: resource_attribute
            name: k8s.pod.uid
        extract:
          metadata:
            - deployment
            - cluster
            - namespace
            - node
            - startTime
          annotations: []
          labels: []
        filter:
          node_from_env_var: KUBE_NODE_NAME
```
this is required config for enriching it with k8s metadata. 
most of it is static. Users doesn't need to know what `passthrough` value they need to pick, nor search though this plugin documentation to properly choose what `pod_association` I need to use. 
So I think in this case, "less configuration options -> less confusion" doesn't quite fit here. it is more like "more guided helm chart -> less confusion". 
We are helping users by presenting the configuration they need to look and hiding what they don't need to waste their minds on.

dmitryax commented on charts/opentelemetry-collector/templates/_pod.tpl:
```diff
@@ -37,10 +37,15 @@ containers:
       {{- . | toYaml | nindent 6 }}
       {{- end }}
-    livenessProbe:
+    livenessProbe: 
```
nit: trailing space

tigrannajaryan commented on charts/opentelemetry-collector/templates/_config.tpl:
```diff
@@ -131,22 +131,15 @@ service:
-            expr: '$$record matches "^[^ Z]+ "'
-          - output: parser-containerd
-            expr: '$$record matches "^[^ Z]+Z"'
+      {{- if eq .Values.agentCollector.containerLogs.containerRunTime "cri-o" }}
```
I do not fully understand what this change does. Did this change delete the ability to automatically detect the format used by kubernetes and requires the user to specify the format?

sumo-drosiek replied:
It looks like that

tigrannajaryan replied:
@rockb1017 please clarify why want to change this. It appears to me automatic detection is valuable since there is one less thing for users to worry about.

rockb1017 replied:
yes, instead of collector detecting the format, we are allowing users to provide their container runtime. this is how it is done in SCK. Though I am not saying this is definitely better way, but at least it will definitely be better for performance. I would say asking users to specify the container runtime once and have more efficient pipeline is better.

tigrannajaryan replied:
It is a tradeoff between ease of use and performance. Both are valid approach and different users likely will have different preferences. I think what we can do is by default have automatic detection of the format and in addition to that allow the user to specify the format, in which case the detection will be disabled and we will have better performance.

dmitryax commented:
Closing this PR since it's too specific for a particular logs collection use case. Feel free to open it again if you want to address the comments
